grid/lib/SharedVariable.py

```diff
@@ -0,0 +1,338 @@
+import torch
+from torch.autograd import Variable, Function
+import inspect
+import random
+import copy
+
+BASE = 10
+KAPPA = 3  # ~29 bits
+
+# TODO set these intelligently
+PRECISION_INTEGRAL = 2
+PRECISION_FRACTIONAL = 5
+PRECISION = PRECISION_INTEGRAL + PRECISION_FRACTIONAL
+BOUND = BASE**PRECISION
+
+# Q field
+field = 2**41  # < 64 bits
+#Q = 2147483648
+Q_MAXDEGREE = 1
+
+
+def encode(rational, precision_fractional=PRECISION_FRACTIONAL):
+    upscaled = (rational * BASE**precision_fractional).long()
+    upscaled.remainder_(field)
+    return upscaled
+
+
+def decode(field_element, precision_fractional=PRECISION_FRACTIONAL):
+    field_element = field_element.data
+    neg_values = field_element.gt(field)
+    #pos_values = field_element.le(field)
+    #upscaled = field_element*(neg_valuese+pos_values)
+    field_element[neg_values] = field-field_element[neg_values]
+    rational = field_element.float() / BASE**precision_fractional
+    return rational
+
+
+def share(secret):
+    first = torch.LongTensor(secret.shape).random_(field)
+    second = (secret - first) % field
+    return [first, second]
+
+
+def reconstruct(shares):
+    return sum(shares) % field
+
+
+def swap_shares(share, party, interface):
+    if (party == 0):
+        interface.send(share)
```

Comment:
Should probably push that to the next PR though

Reply:
So we need to just pass an interface for sending. Was planning this for the next pr

Reply:
I think we can assume that certain torch methods will be identical across backends, e.g. `send_` and `get_` would likely be the ones you need here and would be generalized.  Esp. since the actual IPFS sending in Grid will be abstracted from the JSON creation in PySyft.

Reply:
So here that would mean share.send_(worker) and share.get_().

Reply:
So we would be hooking pytorch for everything? I was planning on having this code work without having pytorch hooked. This would work by having an interface class which had send and recieve methods. There would be a worker interface and a client interface. They would both wrap the grid methods to provide that interface

Reply:
We'll work through this later.